### PR TITLE
feat: implement media-query-mixin

### LINKED ISSUE #36: Research: prototype MediaQueryPropertyMixin
## Problem

In certain components we apply CSS conditionally based on media queries.
We don't have a unified approach for it, so components handle it differently.

## 1. JS API

Example: `vaadin-date-picker` with [`_fullscreenMediaQuery`](https://github.com/vaadin/vaadin-date-picker/blob/21e7b0b5adb67a81e9874ee6bd552c699d59e889/src/vaadin-date-picker-mixin.html#L94-L101) set from JS which toggles the `fullscreen` state attribute on `vaadin-date-picker-overlay`.

When using this approach, we currently hardcode media query also in CSS ([example](https://github.com/vaadin/vaadin-date-picker/blob/a700c2deaeae7a33567dc956565ae0e4376eb323/theme/lumo/vaadin-date-picker-overlay-styles.html#L39)) which makes overriding it more complex (as it needs to be done in multiple places).

## 2. CSS API

Example: `vaadin-app-layout` with following CSS custom properties:

- [`--vaadin-app-layout-touch-optimized`](https://github.com/vaadin/vaadin-app-layout/blob/19795118d35117880af4199f391996d6d6d1bc7c/src/vaadin-app-layout.html#L26)
- [`--vaadin-app-layout-drawer-overlay`](https://github.com/vaadin/vaadin-app-layout/blob/19795118d35117880af4199f391996d6d6d1bc7c/src/vaadin-app-layout.html#L158)

Both of these are observed in window `resize` event handler (there is no platform way to observe changes to custom CSS properties so far, see also discussion about [`observedStyles`](https://github.com/w3c/webcomponents/issues/856)).

Then those `true` / `false` are used to show / hide parts of the DOM and one of them is synced to `overlay` property (which reflects to attribute and is read-only).

While this approach lets the user to write their CSS and override the defaults, it still requires boilerplate work for every component and that doesn't sound like a good DX.

## Solution proposal

Let's assume we have a mixin with the following API:
```js
static get mediaQueries() {
  return {
    touchOptimized: '(max-width: 800px) and (min-height: 500px)'
  };
}
```

Then we declare property with `noAccessor` flag:
```js
static get properties() {
  return {
    touchOptimized: {
      type: Boolean,
      reflect: true, // optional. set if we need to sync with an attribute
      attribute: 'touch-oprimized' // optional. set to have dash-case
    }
  };
}
```

For each of the properties declared like this, the mixin does the following:

- create `window.matchMedia` and add listener to the `mediaQueries[prop]`
- create custom accessors that make sure the property is read only
- toggle property value between `true` and `false` depending on media query
- when attribute is set manually, keep it and override media query based value
  - this is needed to allow forcing specific appearance e.g. `touch-optimized`
- when attribute is removed manually, watch to media query changes again

This approach (only allow to force the value that is not the default one) is inspired by the way how we handle `dir` attribute in `DirMixin` when set by the user to a different value).

### PR DESCRIPTION
**Pull request by web-padawan**
Fixes #36 

The implementation that we ended up while discussing with @yuriy-fix looks like this:

```ts
@customElement('mq-basic-element')
class MyElement extends MediaQueryMixin(LitElement) {

  @mediaProperty({ media: '(max-width: 420px)' }) 
  fullscreen: boolean | null | undefined;
}
```

TL;DR: `true` when query matches, and `false` otherwise.

- `attribute: false` is forced so we ignore setting attribute
- property is read-only so we ignore manual modifications

### CODE REVIEW

File: packages/media-query-mixin/test/media-query-mixin.test.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+const listeners: { [key: string]: () => void } = {};
+const queryMatches: { [key: string]: boolean } = {};
+
+const runQuery = (query: string, matches: boolean) => {
+  queryMatches[query] = matches;
+  listeners[query]();
+};
+
+const FULLSCREEN = '(max-width: 420px) and (max-heigh: 420px)';
```
Comment (lkraav) [inline packages/media-query-mixin/test/media-query-mixin.test.ts]: `max-heigh` typo missing `t`.

Comment (web-padawan) [reply packages/media-query-mixin/test/media-query-mixin.test.ts]: Thanks!

Comment (web-padawan) [reply packages/media-query-mixin/test/media-query-mixin.test.ts]: Done.

File: packages/media-query-mixin/media-query-mixin.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            media
+          }
+        : options;
+
+      ((base as unknown) as typeof LitElement).createProperty(name, opts);
+
+      if (isMediaString) {
+        const query = window.matchMedia(media as string);
+        this._mediaQueries[prop] = { query };
+        this._queryProps.add(prop);
```
Comment (yuriy-fix) [inline packages/media-query-mixin/media-query-mixin.ts]: Why do we need `_queryProps ` if we can get a reference with `_mediaQueries[prop]`?

File: packages/media-query-mixin/media-query-mixin.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        Object.defineProperty(this.prototype, prop, {
+          get(): boolean | null | undefined {
+            return (this as { [key: string]: boolean | null | undefined })[key];
+          },
+          set(this: MediaQuery, value: boolean | null | undefined) {
+            if (MediaQuery._mediaQueries[prop].updating !== true) {
+              return;
+            }
+            const oldValue = ((this as {}) as { [key: string]: unknown })[prop];
+            ((this as {}) as { [key: string]: unknown })[key] = value;
```
Comment (yuriy-fix) [inline packages/media-query-mixin/media-query-mixin.ts]: Nit: A bit complex to understand from the first sight where we saving what values 😅 Probably needs comment with a link to https://github.com/Polymer/lit-element/blob/41e9fd3b8c18b3070f9df2a2f21915b796b2b298/src/lib/updating-element.ts#L306-L320

File: packages/media-query-mixin/test/media-query-mixin.test.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      expect(element.fullscreen).to.be.true;
+    });
+
+    it('should toggle property value when query does not match', async () => {
+      runQuery(FULLSCREEN, false);
+      await element.updateComplete;
+      expect(element.fullscreen).to.be.false;
+    });
+
+    it('should disallow direct property modification attempt', async () => {
```
Comment (yuriy-fix) [inline packages/media-query-mixin/test/media-query-mixin.test.ts]: Should we consider special case of unsetting the prop, to clean up the media query checks by setting prop to `null`
and add test for it?

File: packages/media-query-mixin/media-query-mixin.ts (hunk 3 above)
Comment (web-padawan) [reply packages/media-query-mixin/media-query-mixin.ts]: Thanks, added.

File: packages/media-query-mixin/media-query-mixin.ts (hunk 2 above)
Comment (web-padawan) [reply packages/media-query-mixin/media-query-mixin.ts]: Removed `_queryProps`.

File: packages/media-query-mixin/test/media-query-mixin.test.ts (hunk 4 above)
Comment (web-padawan) [reply packages/media-query-mixin/test/media-query-mixin.test.ts]: Discussed in person that we want to align it with how `iron-media-query` works.
